src/core/CONVENTIONS.md

```diff
@@ -0,0 +1,212 @@
+}
+
+export interface MyPluginStart {
+  getThing(): Thing;
+}
+
+export interface MyPluginSetupDeps {
+  otherPlugin: OtherPluginSetup;
+  thirdPlugin?: ThirdPluginSetup;  // Optional dependency
+}
+
+export interface MyPluginStartDeps {
+  otherPlugin: OtherPluginStart;
+  thirdPlugin?: ThirdPluginStart;  // Optional dependency
+}
+
+export class MyPlugin implements Plugin<
+  // All of these types are optional. If your plugin does not expose anything
+  // or depend on other plugins, these can be omitted.
+  MyPluginSetup,
+  MyPluginStart,
+  MyPluginSetupDeps,
+  MyPluginStartDeps,
+> {
+
+  public setup(core: CoreSetup, plugins: MyPluginSetupDeps) {
+    // should return a value that matches `MyPluginSetup`
+  }
+
+  public start(core: CoreStart, plugins: MyPluginStartDeps) {
+    // should return a value that matches `MyPluginStart`
+  }
+
+  public stop() { ... }
+}
+```
+
+Difference between `setup` and `start`:
+- `setup` is reserved for "registration" work
+- `start` is where any "running" logic for your plugin would live. This only applies if you need to start listening for
+  outside events (polling for work, listening on a port, etc.)
+
+The bulk of your plugin logic will most likely live inside _handlers_ registered during `setup`.
+
+### Applications
+
+It's important that UI code is not included in the main bundle for your plugin. Our webpack configuration supports
+dynamic async imports to split out imports into a separate bundle. Every app's rendering logic and UI code should
+leverage this pattern.
+
+```tsx
+// my_plugin/public/applications/my_app.ts
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { ApplicationMountContext } from '../../src/core/public';
+
+import { MyAppRoot } from './components/app.ts';
+
+/**
+ * This module will be loaded asynchronously to reduce the bundle size of your plugin's main bundle.
+ */
+export const renderApp = (context: ApplicationMountContext, domElement: HTMLDivElement) => {
+  ReactDOM.render(<MyAppRoot context={context} />, domElement);
+  return () => ReactDOM.unmountComponentAtNode(domElement);
+}
+```
+
+```ts
+// my_plugin/public/plugin.ts
+
+import { Plugin } from '../../src/core/public';
+
+export class MyPlugin implements Plugin {
+  public setup(core) {
+    core.application.register({
+      id: 'my-app',
+      async mount(context, domElement) {
+        // Load application bundle
+        const { renderApp } = await import('./application/my_app');
+        return renderApp(context, domElement);
+      }
+    });
+  }
+}
+```
+
+### Services
+
+Service structure should mirror the plugin lifecycle to make reasoning about how the service is executed more clear.
+
+```ts
+// my_plugin/public/services/my_service.ts
+
+export class MyService {
+  private readonly strings$ = new BehaviorSubject<string[]>();
+
+  public setup() {
+    return {
+      registerStrings: (newString: string) =>
+        this.values$.next([...this.strings$.value, newString]);
```

Comment:
`this.strings$.next([...this.strings$.value, newString]);`

Reply:
If only F2 worked in markdown files 😄